loader/loader.c

```diff
@@ -5127,6 +5127,149 @@ bool loaderGetLayerInterfaceVersion(PFN_vkNegotiateLoaderLayerInterfaceVersion f
     return true;
 }
 
+VKAPI_ATTR VkResult VKAPI_CALL vkLayerCreateDevice(VkInstance instance, VkPhysicalDevice physicalDevice, const VkDeviceCreateInfo *pCreateInfo,
+						   const VkAllocationCallbacks *pAllocator, VkDevice *pDevice, PFN_vkGetInstanceProcAddr layerGIPA, PFN_vkGetDeviceProcAddr *nextGDPA) {
+    VkResult res;
+    VkPhysicalDevice internal_device = VK_NULL_HANDLE;
+    struct loader_device *dev = NULL;
+    struct loader_instance *inst = NULL;
+
+    assert(pCreateInfo->queueCreateInfoCount >= 1);
+
+    if(instance != NULL) {
+      inst = loader_get_instance(instance);
+      internal_device = physicalDevice;
+    } else {
+      struct loader_physical_device_tramp *phys_dev = (struct loader_physical_device_tramp *)physicalDevice;
+      internal_device = phys_dev->phys_dev;
+      inst = (struct loader_instance *)phys_dev->this_instance;
+    }
```

Comment:
I'm not crazy about having this block unwrap a handle based on a null instance. I would recommend having `vkCreateDevice` unwrap the handles outside of this function and pass in the unwrapped handles to this function. That seems simpler than putting the unwrapping logic in here and passing a null instance into this function.